primo/utils/clustering_utils.py

```diff
@@ -145,3 +177,106 @@ def perform_clustering(wd: WellData, distance_threshold: float = 10.0):
     wd.col_names.register_new_columns({"cluster": "Clusters"})
 
     return clustered_data.n_clusters_
+
+
+def perform_louvain_clustering(
+    wd: WellData, distance_threshold: float = 10.0, cluster_threshold: int = 300
+) -> dict:
+    """
+    Partitions the data into smaller clusters using the Louvain community detection method,
+    limiting each well to connect only with its 10 closest neighbors within a threshold distance.
+    Dynamically adjusts the resolution parameter to ensure no cluster exceeds the size threshold.
+
+    Parameters
+    ----------
+    wd : WellData
+        Object containing well data
+
+    distance_threshold : float
+        Threshold distance (in miles) for breaking clusters
+
+    cluster_threshold : int
+        cluster size of the largest cluster
+
+    Returns
+    -------
+    dict
+        Dictionary of lists of wells contained in each cluster
+    """
+    # pylint: disable=too-many-locals
+    existing_clusters = _check_existing_cluster(wd)
+    if existing_clusters is not None:
+        return existing_clusters
+
+    coordinates = list(
+        zip(wd.data[wd.col_names.latitude], wd.data[wd.col_names.longitude])
+    )
+    ball_tree = BallTree(np.radians(coordinates), metric="haversine")
+
+    distances, indices = ball_tree.query(np.radians(coordinates), k=11)
+
+    well_graph = nx.Graph()
+    well_ids = wd.data.index.tolist()
+    well_graph.add_nodes_from(well_ids)
+    # Add edges within distance threshold
+    for i, neighbors in enumerate(indices):
+        well_id_1 = well_ids[i]
+        for j in range(1, len(neighbors)):  # Skip the point itself (index 0)
+            well_id_2 = well_ids[neighbors[j]]
+            if not well_graph.has_edge(well_id_1, well_id_2):
+                distance = distances[i][j] * EARTH_RADIUS  # Convert to distance
+                if distance <= distance_threshold:
+                    well_graph.add_edge(
+                        well_id_1, well_id_2, weight=distance_threshold - distance
+                    )
+
+    # Louvain clustering
+    communities = nx.community.louvain_communities(well_graph, seed=4242, resolution=1)
+    well_cluster_map = {
+        well: cluster_id
+        for cluster_id, community in enumerate(communities)
+        for well in community
+    }
+    wd.add_new_column_ordered(
+        "cluster", "Clusters", [well_cluster_map[well] for well in well_ids]
+    )
+
+    LOGGER.debug("The resolution parameter is set to 1.")
+    LOGGER.debug(
+        f"There are {len(communities)} clusters with the largest cluster containing {max(len(community) for community in communities)} wells."
+    )
+    community_sizes = [len(community) for community in communities]
+    max_cluster_size = max(community_sizes)
+    if max_cluster_size <= cluster_threshold:
+        return _well_clusters(wd)
+
+    resolution = 1.5
```

Comment:
Just for maintainability, I will split this method into two methods. 

- Method 1: Does clustering for a 'fixed' level of resolution
- Method 2: Does a higher level 'approach' where you try different levels of resolution until you get what you need

Reply:
+1

Reply:
Resolved. If have added resolution as a parameter based on the previous comments. The default value for this parameter is -1, in which case we calculate the resolution parameter based on the max cluster size. If a resolution parameter is provided, then we use that value and return the clusters.